sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExpr.atg

```diff
@@ -0,0 +1,396 @@
+//   switch = '$' { digit | letter }. (. Optional semantic action .)
+
+
+// Add comments here.
+// Example for a multi-line block comment:
+COMMENTS FROM "/*" TO "*/" NESTED
+// Example for a single line comment:
+   COMMENTS FROM "//" TO lf
+
+
+// Set the ignored characters (whitespaces) here, the blank character is
+// ignored by default.
+// Example, add line breaks to the ignore set.
+IGNORE cr + lf
+
+
+PRODUCTIONS
+
+// Add your productions here, one must have the same name as the grammar,
+// it will be the start symbol (entry point of your compiler).
+// Example:
+//   Empty = "BEGIN" { Statement } "END".
+//   Statement = ident "=" number { "+" number } .
+DebugExpr		(. LLVMExpressionNode n=null; .)
+=
+Expr<out n> 	(. if(errors.count==0) astRoot =n; .)
+.
+
+PrimExpr<out LLVMExpressionNode n>		(. n=errorObjNode; .)
+=
+ident 								(. n = NF.createVarNode(t.val, scopes);.)
+|
+number 								(. n = NF.createIntegerConstant(Integer.parseInt(t.val)); .)
+|
+floatnumber							(. n = NF.createFloatConstant(Float.parseFloat(t.val)); .)					
+|
+charConst
+|
+stringType
+|
+"(" Expr<out n> ")"
+.
+
+Designator<out LLVMExpressionNode n>	(. LLVMExpressionNode idx=null; List<LLVMExpressionNode> l; .)
+=
+PrimExpr<out n> 
+{
+	 "[" Expr<out idx> "]"
+	|
+	 ActPars<out l>
+	|
+	 "." ident
+	|
+	 "->" ident
+}
+.
+
+ActPars<out List l>		(. LLVMExpressionNode n1=null, n2=null; l = new LinkedList<LLVMExpressionNode>(); .)
+=
+"(" 
+[
+	 Expr<out n1> 		(. l.add(n1); .)
+	 
+	{
+		 "," Expr<out n2> (. l.add(n2); .)
+		
+	}
+	
+]
+")"
+.
+
+UnaryExpr<out LLVMExpressionNode n>		(. n=null; int kind=-1; Pair typeP=null;.)
+=
+Designator<out n>
+|
+UnaryOp<out kind> CastExpr<out n> 		(. switch(kind) {
+												case 0:/*n = address(n)*/ break;
+												case 1: /*deref(n)*/ break;
+												case 2: default: break;
+												case 3: n = NF.createArithmeticOp(ArithmeticOperation.SUB, null,
+												NF.createIntegerConstant(0)
+												, n); break;
+												case 4: /*flip bits*/ break;
+												case 5: /*negate boolean/int*/ break;
+											} .)
+|
+"sizeof" "(" DType<out typeP> ")"	(. n=NF.createSizeofNode((Type)typeP.getLeft()); .)
+.
+
+UnaryOp<out int kind>		(. kind=-1; .)
+=
+"&" 						(. kind=0; .)
+|
+"*" 						(. kind=1; .)
+|
+"+" 						(. kind=2; .)
+|
+"-" 						(. kind=3; .)
+|
+"~" 						(. kind=4; .)
+|
+"!" 						(. kind=5; .)
+.
+CastExpr<out LLVMExpressionNode n>		(. Pair typeP=null; .)
+=
+[
+	IF (IsCast()) "(" DType<out typeP> ")" 
+]
+UnaryExpr<out n>			(. if(typeP!=null) {
+									if((Boolean)typeP.getRight()) {
+										n = NF.createSignedCast(n, (Type) typeP.getLeft());
+									} else {
+										n = NF.createUnsignedCast(n, (Type) typeP.getLeft());
+									}
+										} .)
+.
+
+MultExpr<out LLVMExpressionNode n>		(. LLVMExpressionNode n1=null; .)
+=
+CastExpr<out n> 
+{
+	 "*" CastExpr<out n1> 				(./*null is passed as type, since a type check is done by the arithmetic node anyway*/
```

Comment:
The comment that you use `null` instead of a type because the type check is not necessary should be put in the `NF` rather than this grammar.